bokeh/plotting/helpers.py

```diff
@@ -291,110 +291,6 @@ def _process_tools_arg(plot, tools):
-
-    x_axiscls = _get_axis_class(x_axis_type, plot.x_range)
-    if x_axiscls:
-        if x_axiscls is LogAxis:
-            plot.x_mapper_type = 'log'
-        xaxis = x_axiscls(plot=plot)
-        xaxis.ticker.num_minor_ticks = _get_num_minor_ticks(x_axiscls, x_minor_ticks)
-        axis_label = kw.pop('x_axis_label', None)
-        if axis_label:
-            xaxis.axis_label = axis_label
-        xgrid = Grid(plot=plot, dimension=0, ticker=xaxis.ticker); xgrid
-        if x_axis_location == "above":
-            plot.above.append(xaxis)
-        elif x_axis_location == "below":
-            plot.below.append(xaxis)
-
-    y_axiscls = _get_axis_class(y_axis_type, plot.y_range)
-    if y_axiscls:
-        if y_axiscls is LogAxis:
-            plot.y_mapper_type = 'log'
-        yaxis = y_axiscls(plot=plot)
-        yaxis.ticker.num_minor_ticks = _get_num_minor_ticks(y_axiscls, y_minor_ticks)
-        axis_label = kw.pop('y_axis_label', None)
-        if axis_label:
-            yaxis.axis_label = axis_label
-        ygrid = Grid(plot=plot, dimension=1, ticker=yaxis.ticker); ygrid
-        if y_axis_location == "left":
-            plot.left.append(yaxis)
-        elif y_axis_location == "right":
-            plot.right.append(yaxis)
-
-    border_args = ["min_border", "min_border_top", "min_border_bottom", "min_border_left", "min_border_right"]
-    for arg in border_args:
-        if arg in kw:
-            setattr(plot, arg, kw.pop(arg))
-
-    fill_args = ["background_fill_color", "border_fill_color"]
-    for arg in fill_args:
-        if arg in kw:
-            setattr(plot, arg, kw.pop(arg))
-
-    style_arg_prefix = ["title", "outline"]
-    for prefix in style_arg_prefix:
-        for k in list(kw):
-            if k.startswith(prefix):
-                setattr(plot, k, kw.pop(k))
-
-    if 'toolbar_location' in list(kw):
-        plot.toolbar_location = kw.pop('toolbar_location')
-
-    tool_objs = []
-    temp_tool_str = str()
-
-    if isinstance(tools, list):
-        for tool in tools:
-            if isinstance(tool, Tool):
-                tool_objs.append(tool)
-            elif isinstance(tool, string_types):
-                temp_tool_str+=tool + ','
-            else:
-                raise ValueError("tool should be a string or an instance of Tool class")
-        tools = temp_tool_str
-
-    repeated_tools = []
-
-    for tool in re.split(r"\s*,\s*", tools.strip()):
-        # re.split will return empty strings; ignore them.
-        if tool == "":
-            continue
-
-        tool_obj = _tool_from_string(tool)
-        tool_obj.plot = plot
-
-        tool_objs.append(tool_obj)
-
-    plot.tools.extend(tool_objs)
-
-    for typename, group in itertools.groupby(sorted([ tool.__class__.__name__ for tool in plot.tools ])):
-        if len(list(group)) > 1:
-            repeated_tools.append(typename)
-
-    if repeated_tools:
-        warnings.warn("%s are being repeated" % ",".join(repeated_tools))
-
-    return plot
-
```

Comment:
Here: https://github.com/bokeh/bokeh/pull/4046 I think it makes more sense in that PR anyway, if you want to back out the commit to make this one smaller